script/tool/lib/src/common/plugin_command.dart

```diff
@@ -174,60 +190,82 @@ abstract class PluginCommand extends Command<void> {
-    // Sharding  9 elements into 3 shards should yield shard sizes 3, 3, 3.
-    // Sharding  2 elements into 3 shards should yield shard sizes 1, 1, 0.
+    final List<PackageEnumerationEntry> allPlugins =
+        await _getAllPackages().toList();
+    allPlugins.sort((PackageEnumerationEntry p1, PackageEnumerationEntry p2) =>
+        p1.directory.path.compareTo(p2.directory.path));
     final int shardSize = allPlugins.length ~/ shardCount +
         (allPlugins.length % shardCount == 0 ? 0 : 1);
     final int start = min(shardIndex * shardSize, allPlugins.length);
     final int end = min(start + shardSize, allPlugins.length);
 
-    for (final Directory plugin in allPlugins.sublist(start, end)) {
-      yield plugin;
+    for (final PackageEnumerationEntry plugin
+        in allPlugins.sublist(start, end)) {
+      if (!(filterExcluded && plugin.excluded)) {
+        yield plugin;
+      }
     }
   }
 
-  /// Returns the root Dart package folders of the plugins involved in this
-  /// command execution, assuming there is only one shard.
+  /// Returns the root Dart package folders of the packages involved in this
+  /// command execution, assuming there is only one shard. Depending on the
+  /// command arguments, this may be a user-specified set of packages, the
+  /// set of packages that should be run for a given diff, or all packages.
+  ///
+  /// This will return packages that have been excluded by the --exclude
+  /// parameter, annotated in the entry as excluded.
   ///
-  /// Plugin packages can exist in the following places relative to the packages
+  /// Packages can exist in the following places relative to the packages
   /// directory:
   ///
   /// 1. As a Dart package in a directory which is a direct child of the
-  ///    packages directory. This is a plugin where all of the implementations
-  ///    exist in a single Dart package.
+  ///    packages directory. This is a non-plugin package, or a non-federated
+  ///    plugin.
   /// 2. Several plugin packages may live in a directory which is a direct
   ///    child of the packages directory. This directory groups several Dart
-  ///    packages which implement a single plugin. This directory contains a
-  ///    "client library" package, which declares the API for the plugin, as
-  ///    well as one or more platform-specific implementations.
+  ///    packages which implement a single plugin. This directory contains an
+  ///    "app-facing" package which declares the API for the plugin, a
+  ///    platform interface package which declares the API for implementations,
+  ///    and one or more platform-specific implementation packages.
   /// 3./4. Either of the above, but in a third_party/packages/ directory that
   ///    is a sibling of the packages directory. This is used for a small number
   ///    of packages in the flutter/packages repository.
-  Stream<Directory> _getAllPlugins() async* {
+  Stream<PackageEnumerationEntry> _getAllPackages() async* {
```

Comment:
Here we are calling `PackageEnumerationEntry`s "packages" fwiw.

Reply:
I felt like the type being `PackageEnumerationEntry` in the context of a method whose sole purpose is to enumerate packages was straightforward enough that the ambiguity wouldn't be an issue (especially given that we are following Flutter style, so use explicit types everywhere).